packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts

```diff
@@ -501,5 +506,94 @@ foo?.bar(/* comment */a,
         },
       },
     },
+    {
```

Comment:
These test cases are looking good so far! Can you add some for:

* Combinations of `?.` and `{}`, like the `'(foo1?.foo2 || {}).foo3;'` output below
* `||=` (just to make sure we don't accidentally catch that)
* `'(foo || {}).bar` itself
* A few more types of expressions to the left of the `||`: maybe, a function call, a ternary, one or two others? (making sure edge cases don't sneak in)
* Inside an `if` statement

Reply:
I think it does have some complications with a ternary to the left of `||`
* if the ternary is wrapped with parenthesis the output is wrong as it loses the wrapping parenthesis
(`((foo1 ? foo2 : foo3) || {}).foo4` -> `foo1 ? foo2 : foo3.foo4` which is definitely wrong when `foo2` is falsy), good catch on that.
* if the ternary is not wrapped (`(foo1 ? foo2 : foo3 || {}).foo4`) there are no reported issues which I think is the most correct for most cases (if we report this it definitely has higher false positive rates)

Reply:
Do you think `(foo ?? {}).bar` should also match? It's likely more accurate than `||` in that it only picks `null` or `undefined`, but it's a more modern, same as optional chaining

Reply:
Ha good callout. Yes I think we should handle that -- it's technically closer to `?.`!